bench.py

```diff
@@ -0,0 +1,162 @@
+        vocab = sorted(list(set(data))) if override_vocab is None else override_vocab
+        data_size, vocab_size = len(data), len(vocab)
+        logging.info('data of crop %s has %d characters, vocab of size %d.' % (str(crop), data_size, vocab_size))
+
+        self.stoi = { ch:i for i,ch in enumerate(vocab) }
+        self.itos = { i:ch for i,ch in enumerate(vocab) }
+        self.block_size = block_size
+        self.vocab_size = vocab_size
+        self.data = data
+        self.vocab = vocab
+
+    def __len__(self):
+        return math.ceil(len(self.data) / self.block_size)
+
+    def __getitem__(self, idx):
+        i = np.random.randint(0, len(self.data) - (self.block_size + 1))
+        chunk = self.data[i:i+self.block_size+1]
+        dix = torch.tensor([self.stoi[s] for s in chunk], dtype=torch.long)
+        return dix[:-1], dix[1:]
+
+class CharDataModule(pl.LightningDataModule):
+
+    def __init__(self, batch_size=64, block_size=128, pin_memory=0, num_workers=0):
+        super().__init__()
+        self.batch_size = batch_size
+        self.block_size = block_size
+        self.num_workers = num_workers
+        self.pin_memory = pin_memory
+
+        self.train_dataset = CharDataset('text8', self.block_size, 'train')
+
+    def prepare_data(self): # called only on 1 GPU/machine
+        pass # could technically download text8 here...
+
+    def setup(self, stage): # called for every GPU/machine
+        if stage == 'train' or stage == 'fit':
+            pass # nothing to do, the train_dataset is initialized in the constructor
+        elif stage == 'val':
+            self.val_dataset = CharDataset('text8', self.block_size, 'val', override_vocab=self.train_dataset.vocab)
+        elif stage == 'test':
+            self.test_dataset = CharDataset('text8', self.block_size, 'test', override_vocab=self.train_dataset.vocab)
+        else:
+            raise ValueError(f"stage {stage} is not recognized")
+
+    def train_dataloader(self):
+        loader = DataLoader(self.train_dataset, batch_size=self.batch_size,
+                            shuffle=True, pin_memory=bool(self.pin_memory),
+                            num_workers=self.num_workers)
+        return loader
+
+    def val_dataloader(self):
+        loader = DataLoader(self.val_dataset, batch_size=self.batch_size,
+                            shuffle=False, pin_memory=bool(self.pin_memory),
+                            num_workers=self.num_workers)
+        return loader
+
+    def test_dataloader(self):
+        loader = DataLoader(self.test_dataset, batch_size=self.batch_size,
+                            shuffle=False, pin_memory=bool(self.pin_memory),
+                            num_workers=self.num_workers)
+        return loader
+
+
+# -----------------------------------------------------------------------------
+
+parser = argparse.ArgumentParser()
+parser.add_argument('-x', '--num-epochs', type=int, default=5, help="number of epochs to train for")
```

Comment:
Datamodules can do:

```python
parser = argparse.ArgumentParser()

# enables whatever you have in your init in argparse :) 
parser = CharDataModule.add_argparse_args(parser)

# enable all the trainer flags in argparse
parser = Trainer.add_argparse_args(parser)

args = parser.parse_args()

# now you can init whatever objects automatically as well:
trainer = Trainer.from_argparse_args(args, any_flag_to_override=...)

dm = CharDataModule.from_argparse_args(args)
```

Which lets you do things like:

```bash
python main.py --gpus 2 --num_nodes 3 --batch_size 32

Reply:
Neat! I'll have to read more of the docs